### ISSUE TITLE
Fastparquet reader with degraded performance in dask==2.2

### ISSUE THREAD

**Issue by RodrigoNeves95** [dask/dask#5200]
## Versions

dask==2.2
fastparquet==0.3.1
pyarrow==0.14.1

### Create and save parquet files
```python
import numpy as np
import pandas as pd
import dask.dataframe as dd
import random
import shutil
import fastparquet
import os

#generate dummy df
def generate_df(size, ncols):

    dates = pd.date_range('1980-01-01', periods=size, freq='1T')
    data = np.random.rand(size, ncols)
    
    df = pd.DataFrame(data=data, columns=[str(col) for col in range(ncols)], index=dates)
    return df
  
```
## Generate dask dataframe
```python
small_df = generate_df(1000000, ncols=20)

df1 = dd.from_pandas(small_df, npartitions=40)
if os.path.exists('test_fastparquet'):
    shutil.rmtree('test_fastparquet')
    
df1.to_parquet('test_fastparquet', compression='snappy', engine='fastparquet')

small_df = generate_df(1000000, ncols=20)

df2 = dd.from_pandas(small_df, npartitions=40)
if os.path.exists('test_pyarrow'):
    shutil.rmtree('test_pyarrow')
    
df2.to_parquet('test_pyarrow', compression='snappy', engine='pyarrow')
```

## reading metadata and compute for 20 columns
```python
%%timeit
dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=True)
19.2 s ± 1.91 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=False)
104 ms ± 4.59 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_pyarrow/', engine='pyarrow', gather_statistics=True)
37.8 ms ± 1.02 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', gather_statistics=False)
9.74 ms ± 121 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

ddf1 = dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=True)
ddf2 = dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=False)
ddf3 = dd.read_parquet('test_pyarrow/', engine='pyarrow', gather_statistics=True)
ddf4 = dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', gather_statistics=False)

%%timeit
ddf1.compute()
9.19 s ± 264 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf2.compute()
9.06 s ± 254 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf3.compute()
1.05 s ± 37.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf4.compute()
909 ms ± 52.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

## Generate dask dataframe
```python
small_df = generate_df(1000000, ncols=2)

df1 = dd.from_pandas(small_df, npartitions=40)
if os.path.exists('test_fastparquet'):
    shutil.rmtree('test_fastparquet')
    
df1.to_parquet('test_fastparquet', compression='snappy', engine='fastparquet')

small_df = generate_df(1000000, ncols=2)

df2 = dd.from_pandas(small_df, npartitions=40)
if os.path.exists('test_pyarrow'):
    shutil.rmtree('test_pyarrow')
    
df2.to_parquet('test_pyarrow', compression='snappy', engine='pyarrow')
```
## reading metadata and compute for 2 columns

```python
%%timeit
dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=True)
476 ms ± 13 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=False)
23.3 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_pyarrow/', engine='pyarrow', gather_statistics=True)
22.7 ms ± 3.78 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', gather_statistics=False)
9.16 ms ± 2.18 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

ddf1 = dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=True)
ddf2 = dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=False)
ddf3 = dd.read_parquet('test_pyarrow/', engine='pyarrow', gather_statistics=True)
ddf4 = dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', gather_statistics=False)

%%timeit
ddf1.compute()
2.18 s ± 151 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf2.compute()
2.02 s ± 59.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf3.compute()
326 ms ± 54.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf4.compute()
297 ms ± 15.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

# Now with dask=2.1.0
## Versions

dask==2.1
fastparquet==0.3.1
pyarrow==0.14.1

## Generate dask dataframe
```python
small_df = generate_df(1000000, ncols=20)

df1 = dd.from_pandas(small_df, npartitions=40)
if os.path.exists('test_fastparquet'):
    shutil.rmtree('test_fastparquet')
    
df1.to_parquet('test_fastparquet', compression='snappy', engine='fastparquet')

small_df = generate_df(1000000, ncols=20)

df2 = dd.from_pandas(small_df, npartitions=40)
if os.path.exists('test_pyarrow'):
    shutil.rmtree('test_pyarrow')
    
df2.to_parquet('test_pyarrow', compression='snappy', engine='pyarrow')
```
## reading metadata and compute for 20 columns
```python
%%timeit
dd.read_parquet('test_fastparquet/', engine='fastparquet', infer_divisions=True)
105 ms ± 455 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_fastparquet/', engine='fastparquet', infer_divisions=False)
100 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_pyarrow/', engine='pyarrow', infer_divisions=True)
42 ms ± 8.58 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', infer_divisions=False)

%%timeit
dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', infer_divisions=False)
26.2 ms ± 353 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

ddf1 = dd.read_parquet('test_fastparquet/', engine='fastparquet', infer_divisions=True)
ddf2 = dd.read_parquet('test_fastparquet/', engine='fastparquet', infer_divisions=False)
ddf3 = dd.read_parquet('test_pyarrow/', engine='pyarrow', infer_divisions=True)
ddf4 = dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', infer_divisions=False)

%%timeit
ddf1.compute()
558 ms ± 4.85 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf2.compute()
563 ms ± 12.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf3.compute()
481 ms ± 3.88 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf4.compute()
503 ms ± 23.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

## Generate dask dataframe
```python
small_df = generate_df(1000000, ncols=20)

df1 = dd.from_pandas(small_df, npartitions=40)
if os.path.exists('test_fastparquet'):
    shutil.rmtree('test_fastparquet')
    
df1.to_parquet('test_fastparquet', compression='snappy', engine='fastparquet')

small_df = generate_df(1000000, ncols=20)

df2 = dd.from_pandas(small_df, npartitions=40)
if os.path.exists('test_pyarrow'):
    shutil.rmtree('test_pyarrow')
    
df2.to_parquet('test_pyarrow', compression='snappy', engine='pyarrow')
```
## reading metadata and compute for 2 columns
```python
%%timeit
dd.read_parquet('test_fastparquet/', engine='fastparquet', infer_divisions=True)
20.6 ms ± 784 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_fastparquet/', engine='fastparquet', infer_divisions=False)
19.5 ms ± 933 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_pyarrow/', engine='pyarrow', infer_divisions=True)
24.3 ms ± 2 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', infer_divisions=False)
17.8 ms ± 577 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

ddf1 = dd.read_parquet('test_fastparquet/', engine='fastparquet', infer_divisions=True)
ddf2 = dd.read_parquet('test_fastparquet/', engine='fastparquet', infer_divisions=False)
ddf3 = dd.read_parquet('test_pyarrow/', engine='pyarrow', infer_divisions=True)
ddf4 = dd.read_parquet('test_pyarrow/*.parquet', engine='pyarrow', infer_divisions=False)

%%timeit
ddf1.compute()
124 ms ± 11.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf2.compute()
140 ms ± 19.3 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
ddf3.compute()
267 ms ± 24.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
ddf4.compute()
331 ms ± 42.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

This behaviour is obviously not desired and is most likely related with the last refactor done to parquet engine. I did not go further to isolate the problem since you guys can have a better clue of what is going on. I'm more than happy to help on this btw! 

Note: Everything is running locally on my desktop.

**Comment by martindurant** [comment]
Locally, I am seeing that `.compute()` takes ~380ms for fastparquet, all columns (but I am running without snappy). 

I can confirm that
`dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=True)` is taking a very long time. The profile looks like
```
         60992183 function calls (58035069 primitive calls) in 25.900 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
2960160/3360 7.722    0.000   25.498    0.008 api.py:642(statistics)
  5644800    4.290    0.000    4.290    0.000 {built-in method numpy.frombuffer}
  5644800    2.712    0.000    8.499    0.000 encoding.py:43(read_plain)
   282240    2.706    0.000    4.753    0.000 api.py:699(<listcomp>)
 22579414    2.047    0.000    2.047    0.000 {method 'get' of 'dict' objects}
   134400    1.830    0.000   19.880    0.000 api.py:694(<dictcomp>)
  5644800    1.498    0.000    1.498    0.000 util.py:93(byte_buffer)
  5644800    1.094    0.000    1.713    0.000 util.py:76(ensure_bytes)
  8821295    0.921    0.000    0.922    0.000 {built-in method builtins.isinstance}
        1    0.300    0.300   25.998   25.998 fastparquet.py:158(read_metadata)
  2893155    0.231    0.000    0.231    0.000 {method 'join' of 'str' objects}
     3360    0.180    0.000   25.678    0.008 api.py:171(statistics)
    13480    0.104    0.000    0.160    0.000 api.py:166(<listcomp>)
     6720    0.052    0.000    0.083    0.000 converted_types.py:86(convert)
6812/6810    0.038    0.000    0.038    0.000 {built-in method numpy.array}
   285840    0.038    0.000    0.038    0.000 {built-in method builtins.getattr}
    70560    0.030    0.000    0.038    0.000 schema.py:97(schema_element)
    13480    0.020    0.000    0.182    0.000 api.py:163(columns)
284148/283922    0.019    0.000    0.019    0.000 {built-in method builtins.len}
```

@rjzamora

**Comment by martindurant** [comment]
This is how quickly is *should* go:
```
In [12]: import fastparquet
In [13]: pf = fastparquet.ParquetFile('test_fastparquet/')
In [14]: %timeit pf.statistics
4.66 ms ± 88.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by rjzamora** [comment]
Thanks for raising this @RodrigoNeves95! -  Yep, definitely a bug.  Thanks for finding the fix so quickly @martindurant!

**Comment by martindurant** [comment]
I can also fix this pretty easily on the fastparquet side, and release that, rather than needing a new Dask release. @mrocklin , a preference between where to fix?

**Comment by mrocklin** [comment]
No strong preference.  I think it's reasonable for us to push out a bugfix
Dask release soon if that would be helpful.  It shouldn't take someone more
than an hour.  I would probably wait a week though to let things accrue.

**Comment by martindurant** [comment]
As far as the other measurements go, there are more worries there, but they don't seem to match up with what I am seeing running timeit here.

Note that I (or, preferably, someone else) could maybe also do work on the `fastparquet.api.statistics()` function  to shortcut whenever a row_group has no stats or isn't ordered, or to ignore columns that are not of interest, etc. It'd be a nice little project, with some impact on the many-partition, many-column cases.

**Comment by martindurant** [comment]
Releasing fastparquet with that fix. That will be the last of fastparquet for py2!

**Comment by RodrigoNeves95** [comment]
Thank you @martindurant for the quick fix!

**Comment by RodrigoNeves95** [comment]
> Locally, I am seeing that `.compute()` takes ~380ms for fastparquet, all columns (but I am running without snappy).
> 
> I can confirm that
> `dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=True)` is taking a very long time. The profile looks like
> 
> 
> ```
>          60992183 function calls (58035069 primitive calls) in 25.900 seconds
> 
>    Ordered by: internal time
> 
>    ncalls  tottime  percall  cumtime  percall filename:lineno(function)
> 2960160/3360 7.722    0.000   25.498    0.008 api.py:642(statistics)
>   5644800    4.290    0.000    4.290    0.000 {built-in method numpy.frombuffer}
>   5644800    2.712    0.000    8.499    0.000 encoding.py:43(read_plain)
>    282240    2.706    0.000    4.753    0.000 api.py:699(<listcomp>)
>  22579414    2.047    0.000    2.047    0.000 {method 'get' of 'dict' objects}
>    134400    1.830    0.000   19.880    0.000 api.py:694(<dictcomp>)
>   5644800    1.498    0.000    1.498    0.000 util.py:93(byte_buffer)
>   5644800    1.094    0.000    1.713    0.000 util.py:76(ensure_bytes)
>   8821295    0.921    0.000    0.922    0.000 {built-in method builtins.isinstance}
>         1    0.300    0.300   25.998   25.998 fastparquet.py:158(read_metadata)
>   2893155    0.231    0.000    0.231    0.000 {method 'join' of 'str' objects}
>      3360    0.180    0.000   25.678    0.008 api.py:171(statistics)
>     13480    0.104    0.000    0.160    0.000 api.py:166(<listcomp>)
>      6720    0.052    0.000    0.083    0.000 converted_types.py:86(convert)
> 6812/6810    0.038    0.000    0.038    0.000 {built-in method numpy.array}
>    285840    0.038    0.000    0.038    0.000 {built-in method builtins.getattr}
>     70560    0.030    0.000    0.038    0.000 schema.py:97(schema_element)
>     13480    0.020    0.000    0.182    0.000 api.py:163(columns)
> 284148/283922 0.019    0.000    0.019    0.000 {built-in method builtins.len}
> ```
> 
> @rjzamora

regarding `.compute()` time from what I saw on the dask dashboard, it seemed that the most time lost on doing it was about to serialize and deserialize the data.

**Comment by martindurant** [comment]
Ah, that would be another matter, I didn't realise that you were running on distributed.

For de/serialisation, we could do some work to improve the speed, and such code was in the earlier version of the refactor PR. Specifically, the column_chunk.metadata.encoding_stats are never used anywhere, and all the column_chunk.meta_data.statistics have already been used, and so we won't need them again. I would have to check, but they key-values can probably be removed (all the pandas stuff was already parsed).

**Comment by martindurant** [comment]
With the following rather shonky code, I go from
```
In [4]: df = dd.read_parquet('test_fastparquet/', engine='fastparquet', gather_statistics=False)
In [5]: %timeit len(df)
489 ms ± 152 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
to
```
In [6]: %timeit len(df)
328 ms ± 97.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
on (local) distributed

```diff
--- a/dask/dataframe/io/parquet/fastparquet.py
+++ b/dask/dataframe/io/parquet/fastparquet.py
@@ -292,16 +292,22 @@ class FastParquetEngine(Engine):
         # This is a list of row-group-descriptor dicts, or file-paths
         # if we have a list of files and gather_statistics=False
         if not parts:
-            parts = pf.row_groups
+            partsin = pf.row_groups
+            pf.fmd.key_value_metadata = None
         else:
             pf = None
-        parts = [
-            {
+            partsin = parts
+        parts = []
+        for piece in partsin:
+            if pf is not None:
+                for col in piece.columns:
+                    col.meta_data.statistics = None
+                    col.meta_data.encoding_stats = None
+            part = {
                 "piece": piece,
                 "kwargs": {"pf": pf, "categories": categories_dict or categories},
             }
-            for piece in parts
-        ]
+            parts.append(part)

         return (meta, stats, parts)
```

@rjzamora , maybe you'd like to tidy that up a bit?

**Comment by RodrigoNeves95** [comment]
This is already an awesome improvement, but can you test on your side with dask==2.2.0 and distributed==2.2.0 if you have the same benchmarks as I do?. Because the `%%timeit` benchmarks differences between dask/distributed versions are really weird to me. I suspect that the problem is on the distributed side right now. I'm delving on it right now.

**Comment by martindurant** [comment]
Are you able to determine what the time is being spent on? On a quick check, I find that *submitting* the work is fast (~<30ms), although surprisingly cloudpickle.dumps is being called both via tokenize and in _graph_to_futures; I would expect only the latter. 

Note that when using fastparquet, the first run in a session will involve numba compilation, which may be significant for data of this size.

@mrocklin , would it be appropriate to try to define `__dask_tokenize__` for whatever is taking the time (I expect the row_group objects), or to provide tokens explicitly somewhere?

**Comment by mrocklin** [comment]
Something like __dask_tokenize__ makes sense to me, but I recommend
asking @jcrist who is really the expert on dask protocols.

**Comment by jcrist** [comment]
You could define `__dask_tokenize__`, or register something with `dask.base.normalize_token.register`:

https://docs.dask.org/en/latest/custom-collections.html#deterministic-hashing

I'm not sure where the location of this definition should live, up to you.

> or to provide tokens explicitly somewhere

This would be another option. It's nice because you don't have to define for custom objects, but potentially fragile if the important state to be hashed changes.

**Comment by RodrigoNeves95** [comment]
@martindurant I can do it whenever I have time. Something I did not understand is that if you hot the same benchmarks as I. 

> Are you able to determine what the time is being spent on? On a quick check, I find that _submitting_ the work is fast (~<30ms), although surprisingly cloudpickle.dumps is being called both via tokenize and in _graph_to_futures; I would expect only the latter.
> 
> Note that when using fastparquet, the first run in a session will involve numba compilation, which may be significant for data of this size.
> 
> @mrocklin , would it be appropriate to try to define `__dask_tokenize__` for whatever is taking the time (I expect the row_group objects), or to provide tokens explicitly somewhere?

**Comment by martindurant** [comment]
> if you got the same benchmarks as I did

No, but I skipped snappy, so that will account for at least some differences

**Comment by bryanwweber** [comment]
@RodrigoNeves95 It looks like this was addressed by #5218. If there's still more to do, please feel free to reopen! Thanks 😄
